honeybadger-java/src/main/java/io/honeybadger/reporter/HoneybadgerReporter.java

```diff
@@ -0,0 +1,427 @@
+    }
+
+    @Override
+    public ConfigContext getConfig() {
+        return config;
+    }
+
+    /**
+     * Processes an {@link Iterable} of Strings, discards invalid values and
+     * aggregates all values into an ordered set.
+     *
+     * @param tags tag values
+     * @return unique set of tags in the same order as input values
+     */
+    protected Set<String> aggregateTags(final Iterable<String> tags) {
+        if (tags == null || tags == Collections.EMPTY_SET) {
+            return Collections.emptySet();
+        }
+
+        final Iterator<String> itr = tags.iterator();
+
+        if (!itr.hasNext()) {
+            return Collections.emptySet();
+        }
+
+        final LinkedHashSet<String> tagHashSet = new LinkedHashSet<>();
+
+        while (itr.hasNext()) {
+            final String tag = itr.next();
+
+            if (tag == null) {
+                continue;
+            }
+
+            final String trimmed = tag.trim();
+
+            if (!trimmed.isEmpty()) {
+                tagHashSet.add(trimmed);
+            }
+        }
+
+        return Collections.unmodifiableSet(tagHashSet);
+    }
+
+    @SuppressWarnings("LiteralClassName")
+    protected boolean supportsHttpServletRequest() {
+        try {
+            Class.forName("javax.servlet.http.HttpServletRequest");
+            return true;
+        } catch (ClassNotFoundException e) {
+            return false;
+        }
+    }
+
+    @SuppressWarnings("LiteralClassName")
+    protected boolean supportsPlayHttpRequest() {
+        try {
+            Class.forName("play.mvc.Http", false, this.getClass().getClassLoader());
+            return true;
+        } catch (ClassNotFoundException e) {
+            return false;
+        }
+    }
+
+    protected NoticeReportResult submitError(final Throwable error,
+                                             final io.honeybadger.reporter.dto.Request request,
+                                             final String message,
+                                             final Set<String> tags) {
+        final String errorClassName = error.getClass().getName();
+        if (errorClassName != null &&
+                getConfig().getExcludedClasses().contains(errorClassName)) {
+            return null;
+        }
+
+        final Notice notice = new Notice(getConfig());
+
+        if (request != null) {
+            final String reportedMessage;
+            if (message != null && !message.isEmpty()) {
+                reportedMessage = message;
+            } else {
+                reportedMessage = parseMessage(error);
+            }
+
+            NoticeDetails noticeDetails = new NoticeDetails(
+                    getConfig(), error, tags, reportedMessage);
+            notice.setRequest(request).setError(noticeDetails);
+        } else {
+            NoticeDetails noticeDetails = new NoticeDetails(getConfig(), error, tags);
+            notice.setError(noticeDetails);
+        }
+
+        String json;
```

Comment:
I would put a comment here indicating that we are writing our JSON payload to a string because it allows us to resend it if the PUT HTTP call fails. If we were to use a stream, we wouldn't be able to easily recreate the stream in order to try again. If we really wanted to make it memory efficient, we could pass around a lambda function that would do the stream creation from the `Notice` instance.

Reply:
I think the subsequent retry logic is probably obvious enough, but I'll go ahead and add that.